File: DSA/HR/sparse_arrays.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

class Result {
    public static List<Integer> matchingStrings(List<String> strings, List<String> queries) {
        HashMap<String, Integer> freq = new HashMap<>();

        for (String str : strings) {
            if (freq.containsKey(str)) {
                freq.put(str, freq.get(str) + 1);
            } else {
                freq.put(str, 1);
            }
        }

        List<Integer> output = new ArrayList<>();
        for (String q : queries) {
            output.add(freq.getOrDefault(q, 0));
        }
        return output;
    }

}

File: DSA/HR/sparse_arrays.cpp
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

vector<int> matchingStrings(vector<string> strings, vector<string> queries)
{
    std::vector<int> ans{};
    std::unordered_map<std::string_view, int> counter{};

    for (const std::string &str : strings)
    {
        counter[str]++;
    }

    for (const std::string &q : queries)
    {
        ans.push_back(counter[q]);
    }

    return ans;
}